### PR TITLE
Superapp integration

### PR DESCRIPTION
**Pull request by oschleic**

### CODE REVIEW

File: contracts/PaymentReceiver.sol [hunk 1]
```diff
@@ -1,103 +1,111 @@
...
+    /*
+    * ------------------------------------------------------------
+    * External view functions
+    * ------------------------------------------------------------
+    */
+
+    function gatesOwnedBy(address _owner) external view returns (address[] memory result){
+        result = new address[](addressToGate[_owner].length);
+        for(uint i = 0; i < addressToGate[_owner].length; i++){
+            result[i] = address(addressToGate[_owner][i]);
```
Comment (Lewington-pitsos) [inline contracts/PaymentReceiver.sol:98]: since `addressToGate` is `storage` I think we save gas by storing `addressToGate[_owner]` as a temporary `memory` variable here.

File: contracts/PaymentReceiver.sol [hunk 2]
```diff
@@ -1,103 +1,111 @@
...
             abi.encodeWithSelector(
-                cfa.deleteFlowByOperator.selector,
-                _token,
-                _from,
-                _to,
+                cfa.createFlowByOperator.selector,
+                token,
+                msg.sender,
+                address(gate),
+                flowRate,
```
Comment (Lewington-pitsos) [inline contracts/PaymentReceiver.sol:56]: I think we can use plain old `createFlow` here, which has the advantage of not having to involve the whole "operator" concept (which is a pain in the neck because you need to authorize operators in a separate step).

File: contracts/PaymentReceiver.sol [hunk 3]
```diff
@@ -1,103 +1,111 @@
...
+        return address(newGate);
     }
 
-    function _deleteFlow(address _from, address _to, ISuperToken _token) internal {
+    // Remove gate, calls function in SuperGate that will selfdestruct the conract
+    /*
+    function deleteGate(address _gate) external {
+        
+    }
+    */
```
Comment (Lewington-pitsos) [inline contracts/PaymentReceiver.sol]: can we kill these comments?

File: contracts/SuperGate.sol [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+        if(inFlowRate == int96(0)){
+            newCtx = cfaV1.deleteFlowWithCtx(newCtx, address(this), owner, acceptedToken);
+        }
+        else if(outFlowRate != int96(0)){
+            newCtx = cfaV1.updateFlowWithCtx(ctx, owner, acceptedToken, inFlowRate);
+        }
+        else{
+            newCtx = cfaV1.createFlowWithCtx(ctx, owner, acceptedToken, inFlowRate);
+        }
```
Comment (Lewington-pitsos) [inline contracts/SuperGate.sol:163]: these two cases are identical

File: contracts/SuperGate.sol [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        // for each callback which will revert.
+        // Developers will want to avoid reverting in Super App callbacks, 
+        // In particular, you want to avoid reverting within the termination callback
+        // (see rules below regarding the termination callback for more info)
+        // you need to make sure only those actually implemented (overridden)
+        // are ever invoked. That's achieved by setting the _NOOP flag for those
+        // callbacks which you don't need and didn't implement.
+        uint256 configWord =
+            SuperAppDefinitions.APP_LEVEL_FINAL |
+            SuperAppDefinitions.AFTER_AGREEMENT_UPDATED_NOOP;
```
Comment (Lewington-pitsos) [inline contracts/SuperGate.sol:79]: we actually have this callback defined below, it shouldn't be registered as a noop

**Review by Lewington-pitsos — CHANGES_REQUESTED** [review]
I added a few commits which add a test for an edge use case: when the user updates a flow to a SuperGate. This seems to cause a very strange gas related error, which I assumed had nothing to do with the new code until another test which merely updates a flow normally passed without issue.

We can probably skip this test for now, but do give it a run and see if any of it makes sense. The impression I get from google is that that error is commonly thrown in many different  reversion cases.
